added some helpers to make it easier to open or create a pail

EricGebhart opened the pull request:
These are 3 functions I use a lot. Find-or-create is especially nice in practice. I noticed that  I could just use create with fail-on-exists set to false, but it seemed clearer to try opening the pail first.

It could be that find-or-create should take the optional arguments for create as well but it didn't seem as important.

I wasn't sure how to begin to test them.

EricGebhart committed:
Added find-or-create to make it easier to open or create a pail.
Changed create to take a PailSpec or PailStructure.
Added write-objects to make it easy to write a list of objects.

dcuddeback commented on src/main/clojure/clj_pail/core.clj:
```diff
@@ -25,6 +25,25 @@
+  "Get a pail from a path, or create one if not found"
+  (let [pail (try (pail path)
+             (catch Exception e (create-pail structure path)))]
+    pail))
```
There's no need for a `let` expression here. You can return the result of the try/catch directly.

EricGebhart replied:
I see that now that you pointed it out. funny what we see and don't see.

----- David Cuddeback's Original Message -----

> > @@ -25,6 +25,25 @@
> >      (Pail/create path spec fail-on-exists)))
> > 
> > +(defn create-pail [structure path & {:as create-key-args}]
> > -  "Create a pail from a PailStructure at 'path'."
> > -  (let [spec (spec structure)
> > -        pail (apply create spec path (mapcat identity create-key-args))]
> > -    pail))
> >   +
> >   +(defn find-or-create [structure path]
> > -  "Get a pail from a path, or create one if not found"
> > -  (let [pail (try (pail path)
> > -             (catch Exception e (create-pail structure path)))]
> > -    pail))
> 
> There's no need for a `let` expression here. You can return the result of the try/catch directly.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1/files#r8629041

## 

Eric Gebhart
e.a.gebhart@gmail.com

dcuddeback commented on src/main/clojure/clj_pail/core.clj:
```diff
@@ -25,6 +25,25 @@
+             (catch Exception e (create-pail structure path)))]
+    pail))
+
+(defn object-write
```
I'm curious what you use this function for. HDFS is designed for reading and writing a lot of data at once. If I recall correctly, any time you open a pail for writing, it creates a new file. HDFS performance degrades rapidly when you have lots of small files. I'm afraid that this function would invite people to use HDFS in a way that doesn't scale, but if there's a legitimate use case for it, I'd be okay with keeping it and adding a warning to the documentation.

dcuddeback commented on src/main/clojure/clj_pail/core.clj:
```diff
@@ -25,6 +25,25 @@
+  [pail object]
+  (let [os (.openWrite pail)]
+    (.writeObject os object)
+    (.close os)))
```
Clojure's `with-open` is an exception-safe version of the `(let [x ...] ... (.close x))` pattern: http://clojure.github.io/clojure/clojure.core-api.html#clojure.core/with-open

EricGebhart replied:
I think it's a matter of me still learning this.  I haven't been
working with a lot of data. While I thought there was a better way,
I haven't spent any time on it. It looks inefficient to me, but it's
what I've been doing.

----- David Cuddeback's Original Message -----

> > -  (let [spec (spec structure)
> > -        pail (apply create spec path (mapcat identity create-key-args))]
> > -    pail))
> >   +
> >   +(defn find-or-create [structure path]
> > -  "Get a pail from a path, or create one if not found"
> > -  (let [pail (try (pail path)
> > -             (catch Exception e (create-pail structure path)))]
> > -    pail))
> >   +
> >   +(defn object-write
> > -  "Open pail and write an object to it"
> > -  [pail object]
> > -  (let [os (.openWrite pail)]
> > -    (.writeObject os object)
> > -    (.close os)))
> 
> Clojure's `with-open` is an exception-safe version of the `(let [x ...] ... (.close x))` pattern: http://clojure.github.io/clojure/clojure.core-api.html#clojure.core/with-open
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1/files#r8629093

## 

Eric Gebhart
e.a.gebhart@gmail.com

EricGebhart replied:
I had just started looking at the object write function again a couple
of days ago. I wanted the convenience of it. I used open-with in a
couple of different places, but didn't get a working generic function out of it.
I didn't spend too much time on it. I realize now that it was laziness
that kept it from working in the function I made.

So how about this one. It is convenient and it encourages efficient and
idiomatic use. And it doesn't suffer from laziness.

It's sort of the opposite of object-seq.

(defn write-objects
  "Write a list of objects to a pail"
  [pail objects](with-open [writer %28.openWrite pail%29]
    %28doall %28map #%28.writeObject writer %%29 objects%29%29))

Eric

----- David Cuddeback's Original Message -----

> > -  (let [spec (spec structure)
> > -        pail (apply create spec path (mapcat identity create-key-args))]
> > -    pail))
> >   +
> >   +(defn find-or-create [structure path]
> > -  "Get a pail from a path, or create one if not found"
> > -  (let [pail (try (pail path)
> > -             (catch Exception e (create-pail structure path)))]
> > -    pail))
> >   +
> >   +(defn object-write
> > -  "Open pail and write an object to it"
> > -  [pail object]
> > -  (let [os (.openWrite pail)]
> > -    (.writeObject os object)
> > -    (.close os)))
> 
> Clojure's `with-open` is an exception-safe version of the `(let [x ...] ... (.close x))` pattern: http://clojure.github.io/clojure/clojure.core-api.html#clojure.core/with-open
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1/files#r8629093

## 

Eric Gebhart
e.a.gebhart@gmail.com

dcuddeback commented on src/main/clojure/clj_pail/core.clj:
```diff
@@ -25,6 +25,25 @@
     (Pail/create path spec fail-on-exists)))
 
 
+(defn create-pail [structure path & {:as create-key-args}]
```
I like this, because it saves people from having to manually create a `PailSpec`. We could merge this with the existing `create` function so that it works with `PailStructure` and `PailSpec` objects interchangeably.

dcuddeback replied:
I just noticed that `find-or-create` uses `create-pail` as it's fallback. If we merge the functionality into `create`, then `find-or-create` would work with `PailSpec` objects as well. I'm starting to lean toward merging the functionality into one function.

EricGebhart replied:
I like that. I created it to start just to make it easier to open a pail
and not worry about if it was already there or not.  But I didn't want
to take away the granularity you already had in the other functions.

Eric

----- David Cuddeback's Original Message -----

> > @@ -25,6 +25,25 @@
> >      (Pail/create path spec fail-on-exists)))
> > 
> > +(defn create-pail [structure path & {:as create-key-args}]
> 
> I just noticed that `find-or-create` uses `create-pail` as it's fallback. If we merge the functionality into `create`, then `find-or-create` would work with `PailSpec` objects as well. I'm starting to lean toward merging the functionality into one function.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1/files#r8629141

## 

Eric Gebhart
e.a.gebhart@gmail.com

EricGebhart replied:
I like that idea. 

So we would have create take a pailstructure instead of a spec.

pail-create goes away.

Then add the {:as create-key-args} to find-or-create...

Is that what you are thinking?

----- David Cuddeback's Original Message -----

> > @@ -25,6 +25,25 @@
> >      (Pail/create path spec fail-on-exists)))
> > 
> > +(defn create-pail [structure path & {:as create-key-args}]
> 
> I just noticed that `find-or-create` uses `create-pail` as it's fallback. If we merge the functionality into `create`, then `find-or-create` would work with `PailSpec` objects as well. I'm starting to lean toward merging the functionality into one function.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1/files#r8629141

dcuddeback commented:
Thanks for sending this in. I really like the idea of `create-pail` from a `PailStructure` and `find-or-create`. I'd like to know more about the use case for `object-write`.

EricGebhart commented:
This version is looking a lot cleaner.  Let me know what you think about write-objects. It's nice for small stuff and it also encourages good practice.  I removed create-pail and adapted create to take Pail Structure. find-or-create takes the optional create args and passes them on.  I also removed the let.

dcuddeback commented:
I like `write-objects`. I was thinking the same thing. If you still want to write a single object, you can do it by wrapping that object as a vector: `(write-objects pail [object])`, but it can also be used to write a lazy `seq` of objects. It's much more general-purpose than `object-write` was.

It looks like you abandoned being able to create a Pail from a `PailStructure` or `PailSpec`. Were there problems with this approach? I wish there was a way for me to attach a patch to this PR, so that I could make a suggestion. I was thinking about something along these lines for the `create` function:

``` clojure
(defn ^Pail create
  "Creates a Pail from a PailSpec at `path`."
  [spec-or-structure path & {:keys [filesystem fail-on-exists]
                             :or {fail-on-exists true}}]
  (if (instance? spec-or-structure PailStructure)
    (recur (spec spec-or-structure))
    (if filesystem
      (Pail/create filesystem path spec-or-structure fail-on-exists)
      (Pail/create path spec-or-structure fail-on-exists))))
```

The `^PailSpec` that's in the original function is a type hint, which tells Clojure that the parameter will always be of that type, so that the compiler can avoid unnecessary reflection. That has to be removed to be able to work with `PailSpec` or `PailStructure` interchangeably.

dcuddeback commented on src/main/clojure/clj_pail/core.clj:
```diff
@@ -24,6 +24,19 @@
+  "Write a list of objects to a pail"
+  [pail objects]
+  (with-open [writer (.openWrite pail)]
+    (doall (map #(.writeObject writer %) objects))))
```
Usually `map` implies that you care about the result. In fact, `doall` will retain the head of the seq, so if this is used to write a lazy seq containing gigabytes of objects, this will consume gigabytes of memory, returning a list of `nil` objects. I think what we want here is `doseq`, because we only care about the side-effects and `doseq` doesn't retain the head:

``` clojure
(doseq [o objects]
  (.writeObject writer o))
```

EricGebhart replied:
That makes perfect sense. 

----- David Cuddeback's Original Message -----

> > @@ -24,6 +24,19 @@
> >      (Pail/create filesystem path spec fail-on-exists)
> >      (Pail/create path spec fail-on-exists)))
> > 
> > +(defn find-or-create [pstruct path & {:as create-key-args}]
> > -  "Get a pail from a path, or create one if not found"
> > -  (try (pail path)
> > -       (catch Exception e
> > -         (apply create (spec pstruct) path
> > -                (mapcat identity create-key-args)))))
> >   +
> >   +(defn write-objects
> > -  "Write a list of objects to a pail"
> > -  [pail objects]
> > -  (with-open [writer (.openWrite pail)]
> > -    (doall (map #(.writeObject writer %) objects))))
> 
> Usually `map` implies that you care about the result. In fact, `doall` will retain the head of the seq, so if this is used to write a lazy seq containing gigabytes of objects, this will consume gigabytes of memory, returning a list of `nil` objects. I think what we want here is `doseq`, because we only care about the side-effects and `doseq` doesn't retain the head:
> 
> ``` clojure
> (doseq [o objects]
>   (.writeObject writer o))
> ```
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1/files#r8665330

## 

Eric Gebhart
e.a.gebhart@gmail.com

dcuddeback commented:
Come to think of it, `Pail/create` (the Java method) will return an existing pail if the `failOnExists` flag is `false`. https://github.com/nathanmarz/dfs-datastores/blob/b58287eaf31f0ded5a993a9c44cc76f5c2b67fb8/dfs-datastores/src/main/java/com/backtype/hadoop/pail/Pail.java#L159

If I'm reading the code correctly, `find-or-create` is functionally equivalent to calling `create` with the `:fail-on-exists` set to `false`. However, that's not very clear from reading the code or the documentation. What I like about `find-or-create` is that it says very clearly what it's expected behavior should be.

We could have `find-or-create` pass it's arguments through to `create` with the `:fail-on-exists` flag overridden to be `false`. `Pail/create` does some additional checking to see if the `PailSpec` matches the one at the given path.

EricGebhart commented:
I only abandoned it because I had changed create to take a PailStructure
instead of a pail spec. And backwards compatibility was broken.  Your
version addresses that. And makes create easier to use.

----- David Cuddeback's Original Message -----

> I like `write-objects`. I was thinking the same thing. If you still want to write a single object, you can do it by wrapping that object as a vector: `(write-objects pail [object])`, but it can also be used to write a lazy `seq` of objects. It's much more general-purpose than `object-write` was.
> 
> It looks like you abandoned being able to create a Pail from a `PailStructure` or `PailSpec`. Were there problems with this approach? I wish there was a way for me to attach a patch to this PR, so that I could make a suggestion. I was thinking about something along these lines for the `create` function:
> 
> ``` clojure
> (defn ^Pail create
>   "Creates a Pail from a PailSpec at `path`."
>   [spec-or-structure path & {:keys [filesystem fail-on-exists]
>                              :or {fail-on-exists true}}]
>   (if (instance? spec-or-structure PailStructure)
>     (recur (spec spec-or-structure))
>     (if filesystem
>       (Pail/create filesystem path spec-or-structure fail-on-exists)
>       (Pail/create path spec-or-structure fail-on-exists))))
> ```
> 
> The `^PailSpec` that's in the original function is a type hint, which tells Clojure that the parameter will always be of that type, so that the compiler can avoid unnecessary reflection. That has to be removed to be able to work with `PailSpec` or `PailStructure` interchangeably.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1#issuecomment-31650262

## 

Eric Gebhart
e.a.gebhart@gmail.com

EricGebhart commented:
I had actually been doing that interchangebly with find-or-create.

Something like this.

(def pail-connection (-> (DataUnitPailStructure.)
                         (pl/spec)
                         (pl/create "example_output" :fail-on-exists false)))

It appears to do the same thing as find-or-create and reading the java code, as far as I
did, confirmed it.

I like find-or-create for the same reasons you do. It's very clear that
it's going to try to open an existing pail and then create one if it
doesn't exist.

At the same time I could see doing it with just create and
:fail-on-exists false. It's less complicated and calling it find-or-create
clarifies it's use. 

(defn find-or-create [structure path](-> %28spec structure%29
           %28create path :fail-on-exists false%29))

Although passing the keyword args on to create is suddenly more complicated and a bit 
confusing.  Where create args actually works in an understandable and predictable way 
with the current version of find-or-create.  One thing I liked is
that by default, find-or-create will fail if it can't find a pail and the directory
exists.

I think changing Create to take a pail structure or a spec  along with
the try, catch version of find-or-create gives the most flexibility and
usefulness. They are also very clear about how they will behave with
various options.

Although I do like the idea of checking the PailSpec against the existing pail.

Eric

----- David Cuddeback's Original Message -----

> Come to think of it, `Pail/create` (the Java method) will return an existing pail if the `failOnExists` flag is `false`. https://github.com/nathanmarz/dfs-datastores/blob/b58287eaf31f0ded5a993a9c44cc76f5c2b67fb8/dfs-datastores/src/main/java/com/backtype/hadoop/pail/Pail.java#L159
> 
> If I'm reading the code correctly, `find-or-create` is functionally equivalent to calling `create` with the `:fail-on-exists` set to `false`. However, that's not very clear from reading the code or the documentation. What I like about `find-or-create` is that it says very clearly what it's expected behavior should be.
> 
> We could have `find-or-create` pass it's arguments through to `create` with the `:fail-on-exists` flag overridden to be `false`. `Pail/create` does some additional checking to see if the `PailSpec` matches the one at the given path.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1#issuecomment-31651912

## 

Eric Gebhart
e.a.gebhart@gmail.com

dcuddeback commented:
I updated the code so that `find-or-create` delegates to `create` with `:fail-on-exists false`. I also fixed the handling of `PailStructure`s in `create`. The code that I attached earlier in this thread was wrong. I didn't test it before I wrote it in here. :D 

Let me know what you think about the latest version. I'm not sure what you meant by, "Although passing the keyword args on to create is suddenly more complicated and a bit confusing.  Where create args actually works in an understandable and predictable way with the current version of find-or-create." Does the current version address your concern?

If you like it, I'm ready to merge it in. You can feel free to squash your commits if you want.

EricGebhart commented:
I was thinking that the way it was was better because it gets confusing
if we set fail-on-exists inside find-or-create.

I had a case where I wanted to open a pail, but the pail wasn't where I
thought it was. find-or-create with the try - catch would fail in that
instance.  Unless I told it not to.

This find-or-create would just create a new pail.

If I wanted this behavior, I could always pass fail-on-exists false to
find-or-create.  The current version doesn't allow that fine of control
over the behavior.

----- David Cuddeback's Original Message -----

> I updated the code so that `find-or-create` delegates to `create` with `:fail-on-exists false`. I also fixed the handling of `PailStructure`s in `create`. The code that I attached earlier in this thread was wrong. I didn't test it before I wrote it in here. :D 
> 
> Let me know what you think about the latest version. I'm not sure what you meant by, "Although passing the keyword args on to create is suddenly more complicated and a bit confusing.  Where create args actually works in an understandable and predictable way with the current version of find-or-create." Does the current version address your concern?
> 
> If you like it, I'm ready to merge it in. You can feel free to squash your commits if you want.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1#issuecomment-31802709

EricGebhart commented:
You know, on the other hand I can totally see the argument that
find-or-create should do one or the other regardless.

It's not like it's find-or-maybe-create if everything is ok.
:-)

Although I think having a choice is good.

----- David Cuddeback's Original Message -----

> I updated the code so that `find-or-create` delegates to `create` with `:fail-on-exists false`. I also fixed the handling of `PailStructure`s in `create`. The code that I attached earlier in this thread was wrong. I didn't test it before I wrote it in here. :D 
> 
> Let me know what you think about the latest version. I'm not sure what you meant by, "Although passing the keyword args on to create is suddenly more complicated and a bit confusing.  Where create args actually works in an understandable and predictable way with the current version of find-or-create." Does the current version address your concern?
> 
> If you like it, I'm ready to merge it in. You can feel free to squash your commits if you want.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1#issuecomment-31802709

dcuddeback commented:
This version of `find-or-create` opens an existing pail or creates a new one if the pail doesn't exist. The difference between this and the other version is that the other version will open the pail even if the `PailSpec` doesn't match. This one throws an exception if you try to open an existing pail, but the `PailSpec` doesn't match the one you pass as an argument.

Here are the scenarios that I tried. The `try/catch` column refers to trying to open the pail with the `Pail` constructor, then creating it if that fails. The `create` column refers to delegating to the `create` function with `:fail-on-exists` set to `false`.

```
+------------------+-----------+-----------+
| Scenario         | try/catch | create    |
+------------------+-----------+-----------+
| no directory     | creates   | creates   |
| empty directory  | creates   | creates   |
| wrong PailSpec   | opens     | exception |
| correct PailSpec | opens     | opens     |
+------------------+-----------+-----------+
```

Is there another scenario that I'm overlooking?

EricGebhart commented:
Only one. :-)  And probably the least important one.

I like that the current version fails with the wrong pailspec 
more than I care about this scenario..

> +--------------------+-----------+-----------+
> | Scenario           | try/catch | create    |
> +--------------------+-----------+-----------+
> | directory, no pail | fails     | creates   |
> +--------------------+-----------+-----------+

Eric

----- David Cuddeback's Original Message -----

> This version of `find-or-create` opens an existing pail or creates a new one if the pail doesn't exist. The difference between this and the other version is that the other version will open the pail even if the `PailSpec` doesn't match. This one throws an exception if you try to open an existing pail, but the `PailSpec` doesn't match the one you pass as an argument.
> 
> Here are the scenarios that I tried. The `try/catch` column refers to trying to open the pail with the `Pail` constructor, then creating it if that fails. The `create` column refers to delegating to the `create` function with `:fail-on-exists` set to `false`.
> 
> ```
> +------------------+-----------+-----------+
> | Scenario         | try/catch | create    |
> +------------------+-----------+-----------+
> | no directory     | creates   | creates   |
> | empty directory  | creates   | creates   |
> | wrong PailSpec   | opens     | exception |
> | correct PailSpec | opens     | opens     |
> +------------------+-----------+-----------+
> ```
> 
> Is there another scenario that I'm overlooking?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1#issuecomment-31999606

## 

Eric Gebhart
e.a.gebhart@gmail.com

dcuddeback commented:
Isn't that the same as my empty directory scenario?

EricGebhart commented:
I didn't think it was.  try/catch didn't specify fail-on-exist false so
if the directory exists it would fail.

I thought I might be crazy for a minute, so I went and tested it.
It definitely fails if there is a directory but no pail.

IllegalArgumentException Path fresh-pail already exists
com.backtype.hadoop.pail.Pail.create (Pail.java:157)

I think it's a minor thing. And I'm not sure how we could get the
pail/pailspec check and that behavior at the same time.

It's kind of screwy that create would complain about an existing pail if
we didn't specify fail-on-exist false. Yet if we do, it verifies the
spec and opens it. I spent a little bit of time in the java code trying
to see another way and didn't find one.

----- David Cuddeback's Original Message -----

> Isn't that the same as my empty directory scenario?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1#issuecomment-32000201

## 

Eric Gebhart
e.a.gebhart@gmail.com

dcuddeback commented:
You're right. It does fail if an empty directory exists at the path.

EricGebhart commented:
I'm not sure how important that is verses opening a pail that doesn't
match a spec.

----- David Cuddeback's Original Message -----

> You're right. It does fail if an empty directory exists at the path.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1#issuecomment-32001356

dcuddeback commented:
Everything I see in Pail.java that could get us access to the `PailSpec` is private, so I'm not sure how we could have it fail on open and check the `PailSpec`.

EricGebhart commented:
I didn't see a way either. How do you feel about it?

It seems checking the spec is more important.

----- David Cuddeback's Original Message -----

> Everything I see in Pail.java that could get us access to the `PailSpec` is private, so I'm not sure how we could have it fail on open and check the `PailSpec`.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1#issuecomment-32001480

dcuddeback commented:
I could go either way on it.

On one hand, yeah, I think checking the spec is important. Returning a pail with the wrong spec is unexpected and can cause errors later in the code in an unrelated call site. Those are usually pretty annoying to debug.

On the other hand, the existing function for opening a pail doesn't check the spec. So I'm tempted  to say we should use your try/catch method, and then it will gain the functionality of checking the spec if we ever add it to the function that opens a pail.

Which version is more useful to you on your projects?

EricGebhart commented:
I could go either way too.  It does make sense to add the spec check
to the pail function if we can. Then we get it all.

But I haven't seen a way to do that yet. Like you said it's all private
functions in there.

But it is a nasty error later on when you've got a  spec mismatch on a
pail.

I think for me, for now, mismatched pails are not a problem.

----- David Cuddeback's Original Message -----

> I could go either way on it.
> 
> On one hand, yeah, I think checking the spec is important. Returning a pail with the wrong spec is unexpected and can cause errors later in the code in an unrelated call site. Those are usually pretty annoying to debug.
> 
> On the other hand, the existing function for opening a pail doesn't check the spec. So I'm tempted  to say we should use your try/catch method, and then it will gain the functionality of checking the spec if we ever add it to the function that opens a pail.
> 
> Which version is more useful to you on your projects?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1#issuecomment-32001795

dcuddeback commented:
Okay. If you push the try/catch version, I'll merge it in.

dcuddeback commented:
Feel free to squash the commits if you want.

EricGebhart commented:
Ok. I'll take a look around tomorrow and see if I can find a way to verify a pail
to it's spec.

----- David Cuddeback's Original Message -----

> Okay. If you push the try/catch version, I'll merge it in.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1#issuecomment-32002129

dcuddeback commented:
Okay, thanks. Don't worry too much about verifying the `PailSpec`. If you can figure it out, that's great.

EricGebhart commented:
Got them squashed. That was a bit of work with all the back 
and forth we did. Every commit was a conflict.

I think it looks good though.

----- David Cuddeback's Original Message -----

> Feel free to squash the commits if you want.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1#issuecomment-32002137

## 

Eric Gebhart
e.a.gebhart@gmail.com

dcuddeback commented:
Looks good. I'm surprised there were merge conflicts when squashing the commits. I didn't even think that was possible since they're linear patches. Did you use `git rebase -i`?

EricGebhart commented:
Cool.

yep. It was a pain. I've never seen that before.

I must've had at least 6 conflicts.

----- David Cuddeback's Original Message -----

> Looks good. I'm surprised there were merge conflicts when squashing the commits. I didn't even think that was possible since they're linear patches. Did you use `git rebase -i`?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/dcuddeback/clj-pail/pull/1#issuecomment-32098751
